subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/ivy/IvyDynamicRevisionRemoteResolveIntegrationTest.groovy

```diff
@@ -108,6 +108,263 @@ dependencies {
+          
+            int count
+          
+            void supply(ComponentMetadataSupplierDetails details) {
+                def id = details.id
+                println "Providing metadata for \$id"
+                repositoryResourceAccessor.withResource("\${id.group}/\${id.module}/\${id.version}/status.txt") {
+                    details.result.status = new String(it.bytes)
+                }
+                println "Metadata rule call count: \${++count}"
+            }
+          }
+"""
+        when:
+        def projectA1 = ivyHttpRepo.module("group", "projectA", "1.1").publish()
+        def projectA2 = ivyHttpRepo.module("group", "projectA", "1.2").publish()
+        def projectB1 = ivyHttpRepo.module("group", "projectB", "1.1").publish()
+        def projectB2 = ivyHttpRepo.module("group", "projectB", "2.2").publish()
+        ivyHttpRepo.module("group", "projectA", "2.0").publish()
+
+
+        and:
+        expectGetStatusOf(projectB1, 'release')
+        expectGetStatusOf(projectB2, 'integration')
+        expectGetDynamicRevision(projectA2)
+        expectGetDynamicRevision(projectB1)
+
+        then: "custom metadata rule prevented parsing of ivy descriptor"
+        checkResolve "group:projectA:1.+": "group:projectA:1.2", "group:projectB:latest.release": "group:projectB:1.1"
+        outputContains 'Providing metadata for group:projectB:2.2'
+        outputContains 'Providing metadata for group:projectB:1.1'
+        !output.contains('Providing metadata for group:projectA:1.1')
+
+        and: "creates a new instance of rule each time"
+        !output.contains('Metadata rule call count: 2')
+    }
+
+    def "handles errors in a custom metadata provider"() {
+        given:
+        buildFile << """
+          repositories {
+              ivy {
+                  name 'repo'
+                  url '${ivyHttpRepo.uri}'
+                  metadataSupplier(MP)
+              }
+          }
+          
+            configurations { compile }
+            
+            dependencies {
+                compile group: "group", name: "projectA", version: "1.+"
+                compile group: "group", name: "projectB", version: "latest.release"
+            }
+
+          class MP implements ComponentMetadataSupplier {
+          
+            void supply(ComponentMetadataSupplierDetails details) {
+                throw new NullPointerException("meh: error from custom rule")
+            }
+          }
+"""
+        def projectA1 = ivyHttpRepo.module("group", "projectA", "1.1").publish()
+        def projectB1 = ivyHttpRepo.module("group", "projectB", "1.1").publish()
+
+        when:
+        expectListVersions(projectA1)
+        projectA1.ivy.expectGet()
+        expectListVersions(projectB1)
+
+        fails 'checkDeps'
+
+        then:
+
+        errorOutput.contains('Could not resolve group:projectB:latest.release')
+        failure.assertHasCause('meh: error from custom rule')
```

Comment:
More failure coverage:
- Recovers from failure on subsequent execution (i.e. recovers from transient failure)
- Failure reported when requested resource does not exist